### PR TITLE
BLUEBUTTON-1784 Search for patients by ptdcntrct

### PR DESCRIPTION
**Pull request by whytheplatypus**
### Change Details
This creates a search operation that find patients based on their part D contract status for a given month of this year.

### Acceptance Validation
- [x] Integration tests exercising the new feature
- [ ] Readability
- [x] Follows a consistent pattern

### External References

- https://github.com/CMSgov/beneficiary-fhir-data/pull/156 for spec and RFC
- [BLUEBUTTON-1784](https://jira.cms.gov/browse/BLUEBUTTON-1784)

### Security Implications

- [ ] new software dependencies

- [ ] altered security controls

- [x] new data stored or transmitted

- [ ] security checklist is completed for this change

- [ ] requires more information or team discussion to evaluate security implications

### CODE REVIEW

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java [hunk 1]
```diff
@@ -188,6 +190,128 @@ public Bundle searchByLogicalId(
...
+        IncludeIdentifiersMode.determineIncludeIdentifiersMode(requestDetails);
+    List<IBaseResource> patients =
+        matchingBeneficiaries.stream()
+            .map(
+                beneficiary -> {
+                  // Then, null out the HICN and MBI if we're not supposed to be returning those.
+                  if (includeIdentifiersMode != IncludeIdentifiersMode.INCLUDE_HICNS_AND_MBIS) {
+                    beneficiary.setHicnUnhashed(Optional.empty());
+                    beneficiary.setMedicareBeneficiaryId(Optional.empty());
+                  }
```
Comment (whytheplatypus) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java]: This is cribbed from other search operations, is it still practical?

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java [hunk 2]
```diff
@@ -188,6 +190,128 @@ public Bundle searchByLogicalId(
...
+    if (system.equals(CcwCodebookVariable.PTDCNTRCT09.getVariable().getId().toLowerCase()))
+      return CcwCodebookVariable.PTDCNTRCT09;
+    if (system.equals(CcwCodebookVariable.PTDCNTRCT10.getVariable().getId().toLowerCase()))
+      return CcwCodebookVariable.PTDCNTRCT10;
+    if (system.equals(CcwCodebookVariable.PTDCNTRCT11.getVariable().getId().toLowerCase()))
+      return CcwCodebookVariable.PTDCNTRCT11;
+    if (system.equals(CcwCodebookVariable.PTDCNTRCT12.getVariable().getId().toLowerCase()))
+      return CcwCodebookVariable.PTDCNTRCT12;
+    throw new InvalidRequestException("Unsupported identifier system: " + system);
+  }
```
Comment (whytheplatypus) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java]: I'd like a better alternative to this kind of mapping.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java [hunk 3]
```diff
@@ -195,6 +197,153 @@ public Bundle searchByLogicalId(
...
+
+    PagingArguments pagingArgs = new PagingArguments(requestDetails);
+    Bundle bundle =
+        TransformerUtils.createBundle(
+            pagingArgs,
+            "/Patient?",
+            "_has:Coverage.extension",
+            coverageId.getValueAsQueryToken(null),
+            patients);
+    return bundle;
```
Comment (dkrylovsb) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java]: If my reading is correct, paging happens _post-query_, i.e. after all matching benes have been pulled from the DB. Can this become a problem though? Some of our Part D contracts have a lot of benes.

Comment (dkrylovsb) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java]: Bene counts for some of the contracts we have:
image

Comment (whytheplatypus) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java]: +1000 Yes, the paging should absolutely be done by the database. I'll start an attempt at moving them over.

If it turns out to require a more fundamental change than makes sense for this PR then I'll leave this as is and make sure designing the move is in the backlog. For what it's worth the post-db paging is used for all resources (including EoBs) and does work with large data sets.

**Comment by whytheplatypus** [comment]
`apps/bfd-server/dev/api-changelog.md` needs to be updated.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java (hunk 3 above)
Comment (whytheplatypus) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java]: done.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java [hunk 4]
```diff
@@ -195,6 +198,179 @@ public Bundle searchByLogicalId(
...
+            .map(
+                beneficiary -> {
+                  // Null out the unhashed HICNs if we're not supposed to be returning them
+                  if (!hasHICN(includeIdentifiersValues)) {
+                    beneficiary.setHicnUnhashed(Optional.empty());
+                  }
+                  // Null out the unhashed MBIs if we're not supposed to be returning
+                  if (!hasMBI(includeIdentifiersValues)) {
+                    beneficiary.setMedicareBeneficiaryId(Optional.empty());
+                  }
```
Comment (dkrylovsb) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java]: @whytheplatypus We'll need the Patient records coming out of this endpoint to contain un-hashed MBIs (if requested by the `IncludeIdentifiers` param). Based on what I see in the PR, I am pretty sure this is going to be the case, but I'd like to confirm with you anyway. An important item for us to have.

Comment (whytheplatypus) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java]: Yup this work honors the `IncludeIdentifiers` header

Comment (dkrylovsb) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java]: 👍

**Comment by whytheplatypus** [comment]
closing in favor of https://github.com/CMSgov/beneficiary-fhir-data/pull/214

**Pull request closed without merging**